### PR TITLE
Create GitHub workflow to run benchmark suite

### PR DESCRIPTION
**Pull request by MilesCranmer**
GitHub actions are desirable for running benchmarks for a variety of reasons:

1. Consistent across runs (see https://labs.quansight.org/blog/2021/08/github-actions-benchmarks/)
2. Can be run on a schedule or on a trigger (e.g., every new Julia tag), without needing someone to run on their own machine
3. Makes it easy to set up various languages, as there is usually a GitHub action that someone has already set up
4. Lets you change versions of these languages easily

This GitHub action successfully runs a subset of the languages and generates the csv data. The workflow runs for the following languages:
- C, fortran, java, javascript, julia, python, R, and rust.

The following benchmarks are not part of the workflow, for the reasons given below:
- Lua, Go (install fine, but the benchmarks themselves are broken)
- Octave (no maintained github action)
- Mathematica, MATLAB (proprietary)

Current result (it prints the csv at the end):
```
c,iteration_pi_sum,8.028984
c,matrix_multiply,43.012142
c,matrix_statistics,5.007982
c,parse_integers,0.19634
c,print_to_file,20.508051
c,recursion_fibonacci,0.025188
c,recursion_quicksort,0.422955
c,userfunc_mandelbrot,0.08167
fortran,iteration_pi_sum,8.028663
fortran,matrix_multiply,57.163952
fortran,matrix_statistics,8.046266
fortran,parse_integers,0.753935
fortran,print_to_file,113.916496
fortran,recursion_fibonacci,4.4e-5
fortran,recursion_quicksort,0.483927
fortran,userfunc_mandelbrot,7.8e-5
java,iteration_pi_sum,16.370829
java,iteration_sinc_sum,0.049201
java,matrix_multiply,788.083768
java,matrix_statistics,30.276736
java,parse_integers,0.274402
java,print_to_file,99.797282
java,recursion_fibonacci,0.0424
java,recursion_quicksort,1.006608
java,userfunc_mandelbrot,0.136501
javascript,iteration_pi_sum,10.5
javascript,matrix_multiply,2900.0
javascript,matrix_statistics,46.9
javascript,parse_integers,0.64
javascript,print_to_file,118.0
javascript,recursion_fibonacci,0.109
javascript,recursion_quicksort,1.61
javascript,userfunc_mandelbrot,0.149
julia,iteration_pi_sum,8.028063
julia,matrix_multiply,33.387676
julia,matrix_statistics,8.219065
julia,parse_integers,0.137201
julia,print_to_file,18.368588
julia,recursion_fibonacci,0.0482
julia,recursion_quicksort,0.469904
julia,userfunc_mandelbrot,0.0796
python,iteration_pi_sum,630.6591033935547
python,matrix_multiply,49.559593200683594
python,matrix_statistics,51.499128341674805
python,parse_integers,1.6732215881347656
python,print_to_file,54.22806739807129
python,recursion_fibonacci,2.522706985473633
python,recursion_quicksort,11.09170913696289
python,userfunc_mandelbrot,6.908893585205078
r,iteration_pi_sum,236.0
r,matrix_multiply,116.0
r,matrix_statistics,78.0
r,parse_integers,4.0
r,print_to_file,1325.0
r,recursion_fibonacci,10.0
r,recursion_quicksort,22.0
r,userfunc_mandelbrot,20.0
rust,iteration_pi_sum,8.029562
rust,matrix_multiply,46.196557
rust,matrix_statistics,6.52925
rust,parse_integers,0.186271
rust,print_to_file,11.194186
rust,recursion_fibonacci,0.046293
rust,recursion_quicksort,0.428904
rust,userfunc_mandelbrot,0.080522
```

### CODE REVIEW

**Comment by MilesCranmer** [comment]
The Lua benchmark seems out of date as well.

**Comment by MilesCranmer** [comment]
The rest seem to be working. Ready for review.

**Comment by mkitti** [comment]
This os a great contribution. Thank you.

File: .github/workflows/benchmarks.yml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        shell: bash
+    strategy:
+      matrix:
+        os: [ubuntu-latest]
+        java-version: ['17']
+        julia-version: ['1.7.1']
+        python-version: ['3.9']
+        numpy-version: ['1.22']
+        gfortran-version: ['9']  # Note: unused since is built-in.
+        rust-version: ['1.42.0']
```
[the remark this answers is not in the document]
Comment (MilesCranmer) [reply .github/workflows/benchmarks.yml:28]: Thanks. Should I generate the `rust-toolchain` file from within the benchmarks.yml script, given this version?

[the remark this answers is not in the document]
Comment (MilesCranmer) [reply .github/workflows/benchmarks.yml:28]: Actually I'll just put a comment here that it's ignored and we can change later if necessary.

File: .github/workflows/benchmarks.yml [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+          julia -e 'using Pkg; Pkg.add("Compat")'
+      - name: "Set up dSFMT"
+        run: |
+          cd ~/
+          mkdir -p dSFMT
+          cd dSFMT
+          wget https://github.com/MersenneTwister-Lab/dSFMT/archive/refs/tags/v2.2.4.tar.gz
+          echo "39682961ecfba621a98dbb6610b6ae2b7d6add450d4f08d8d4edd0e10abd8174 v2.2.4.tar.gz" | sha256sum --check --status
+          tar -xzf v2.2.4.tar.gz
+          mv dSFMT-*/* ./
```
[the remark this answers is not in the document]
Comment (MilesCranmer) [reply .github/workflows/benchmarks.yml:58]: Actually I can't. This BinaryBuilder version doesn't include the C source code (dSFMT.c), which is required by the benchmark.

It says it's built by the [same version](https://github.com/JuliaBinaryWrappers/dSFMT_jll.jl#sources) though, so I think it's fine?

**Comment by MilesCranmer** [comment]
Ping for review. Let me know if this is ready for merging.

**Comment by MilesCranmer** [comment]
Ping for review. @Enet4 @giordano @mkitti do any of you have write access?

**Comment by mkitti** [comment]
Sorry, I do not have write access.

**Comment by MilesCranmer** [comment]
@StefanKarpinski @johnfgibson would one of you be willing to approve this? Thanks!

File: .github/workflows/benchmarks.yml [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        numpy-version: ['1.22']
+        gfortran-version: ['9']  # Note: unused since is built-in.
+        rust-version: ['1.42.0']  # Note: unused since controlled by `rust/rust-toolchain`
+        js-version: ['16']
+        r-version: ['4.1.2']
+        lua-version: ['2.0.5']  # Note: Not used as benchmark is broken.
+        go-version: ['1.17.4']  # Note: Not used as benchmark is broken.
+    
+    steps:
+      - uses: actions/checkout@v2
```
Comment (DilumAluthge) [inline .github/workflows/benchmarks.yml:35]:
```suggestion
      - uses: actions/checkout@v2
        with:
          persist-credentials: false
```

File: .github/workflows/benchmarks.yml [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    branches:
+      - '*'
+  pull_request:
+    branches:
+      - '*'
+  workflow_dispatch:
+    branches:
+      - '*'
+
+jobs:
```
Comment (DilumAluthge) [inline .github/workflows/benchmarks.yml:14]:
```suggestion
permissions:
  contents: read

jobs:
```

File: .github/workflows/benchmarks.yml [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+on:
+  push:
+    branches:
+      - '*'
+  pull_request:
+    branches:
+      - '*'
+  workflow_dispatch:
+    branches:
+      - '*'
```
Comment (DilumAluthge) [inline .github/workflows/benchmarks.yml:12]:
```suggestion
on:
  pull_request:
  push:
    branches:
      - master
    tags: '*'
  workflow_dispatch:
```

**Review by DilumAluthge — CHANGES_REQUESTED** [review]

File: .github/workflows/benchmarks.yml [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
```
Comment (DilumAluthge) [inline .github/workflows/benchmarks.yml:13]:
```suggestion

concurrency:
  # Skip intermediate builds: all builds except for builds on the `master` or `release-*` branches
  # Cancel intermediate builds: only pull request builds
  group: ${{ github.workflow }}-${{ github.ref }}-${{ github.ref != 'refs/heads/master' || startsWith(github.ref, 'refs/heads/release-') || github.run_number }}
  cancel-in-progress: ${{ startsWith(github.ref, 'refs/pull/') }}

```

**Review by DilumAluthge — CHANGES_REQUESTED** [review]

File: .github/workflows/benchmarks.yml (hunk 3 above)
Comment (MilesCranmer) [reply .github/workflows/benchmarks.yml:35]: Will add. Out of curiosity, what will this change?

**Comment by MilesCranmer** [comment]
Thanks, all suggestions implemented.

File: .github/workflows/benchmarks.yml [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        numpy-version: ['1.22']
+        gfortran-version: ['9']  # Note: unused since is built-in.
+        rust-version: ['1.42.0']  # Note: unused since controlled by `rust/rust-toolchain`
+        js-version: ['16']
+        r-version: ['4.1.2']
+        lua-version: ['2.0.5']  # Note: Not used as benchmark is broken.
+        go-version: ['1.17.4']  # Note: Not used as benchmark is broken.
+    
+    steps:
+      - uses: actions/checkout@v2
```
Comment (giordano) [inline .github/workflows/benchmarks.yml:41]: There is v3 now
```suggestion
      - uses: actions/checkout@v3
```

**Review by DilumAluthge — APPROVED** [review]
Let's give it a try.

**Comment by MilesCranmer** [comment]
Thanks!!

**Comment by mkitti** [comment]
> Thanks

Thank you for your patience and your contribution, @MilesCranmer .

**Comment by ViralBShah** [comment]
The github workflow now will run the whole benchmark and update the plot and show it at:

https://julialang.github.io/Microbenchmarks.jl/dev/

**Pull request merged**